pkg/conversion/workflow.go

```diff
@@ -0,0 +1,129 @@
+package conversion
+
+import (
+	"fmt"
+	"sort"
+
+	timestamppb "google.golang.org/protobuf/types/known/timestamppb"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/tinkerbell/tink/pkg/apis/core/v1alpha1"
+	protoworkflow "github.com/tinkerbell/tink/protos/workflow"
+	"github.com/tinkerbell/tink/workflow"
+)
+
+func WorkflowYamlToCRD(wf *workflow.Workflow, crd *v1alpha1.Workflow) {
+	if wf == nil || crd == nil {
+		return
+	}
+
+	tasks := []v1alpha1.Task{}
+	for _, task := range wf.Tasks {
+		actions := []v1alpha1.Action{}
+		for _, action := range task.Actions {
+			actions = append(actions, v1alpha1.Action{
+				Name:        action.Name,
+				Image:       action.Image,
+				Timeout:     action.Timeout,
+				Command:     action.Command,
+				Volumes:     action.Volumes,
+				Status:      protoworkflow.State_name[int32(protoworkflow.State_STATE_PENDING)],
+				Environment: action.Environment,
+				Pid:         action.Pid,
+			})
+		}
+		tasks = append(tasks, v1alpha1.Task{
+			Name:        task.Name,
+			WorkerAddr:  task.WorkerAddr,
+			Volumes:     task.Volumes,
+			Environment: task.Environment,
+			Actions:     actions,
+		})
+	}
+	crd.Status.GlobalTimeout = int64(wf.GlobalTimeout)
+	crd.Status.Tasks = tasks
+}
+
+func WorkflowCRDToProto(w *v1alpha1.Workflow) *protoworkflow.Workflow {
+	if w == nil {
+		return nil
+	}
+	v, ok := protoworkflow.State_value[w.Status.State]
+	state := protoworkflow.State(v)
+	if !ok {
+		state = protoworkflow.State_STATE_PENDING
+	}
+	return &protoworkflow.Workflow{
+		Id:        w.TinkID(),
+		Template:  w.Spec.TemplateRef,
+		State:     state,
+		CreatedAt: timestamppb.New(w.CreationTimestamp.Time),
+		DeletedAt: metav1ToTimestamppb(w.DeletionTimestamp),
+	}
+}
+
+func WorkflowActionListCRDToProto(wf *v1alpha1.Workflow) *protoworkflow.WorkflowActionList {
+	if wf == nil {
+		return nil
+	}
+	resp := &protoworkflow.WorkflowActionList{
+		ActionList: []*protoworkflow.WorkflowAction{},
+	}
+	for _, task := range wf.Status.Tasks {
+		for _, action := range task.Actions {
+			resp.ActionList = append(resp.ActionList, &protoworkflow.WorkflowAction{
+				TaskName: task.Name,
+				Name:     action.Name,
+				Image:    action.Image,
+				Timeout:  action.Timeout,
+				Command:  action.Command,
+				WorkerId: task.WorkerAddr,
+				Volumes:  append(task.Volumes, action.Volumes...),
+				// TODO: (micahhausler) Dedupe task volume targets overridden in the action volumes?
+				//   Also not sure how Docker handles nested mounts (ex: "/foo:/foo" and "/bar:/foo/bar")
+				Environment: func(env map[string]string) []string {
+					resp := []string{}
+					merged := map[string]string{}
+					for k, v := range env {
+						merged[k] = v
+					}
+					for k, v := range action.Environment {
+						merged[k] = v
+					}
+					for k, v := range merged {
+						resp = append(resp, fmt.Sprintf("%s=%s", k, v))
+					}
+					sort.Strings(resp)
```

Comment:
Is this required or just your pref? Either way :chef_kiss: I approve :D

Reply:
Not strictly required, but makes testing _way_ easier